doc/guides/smf/index.rst

```diff
@@ -0,0 +1,267 @@
+    - e - Function pointer to state's Entry action. Can be NULL.
+    - r - Function pointer to state's Run action. Can be NULL.
+    - x - Function pointer to state's Exit action. Can be NULL.
+    - p - Function pointer to state's Parent state. Can be NULL.
+
+**HSM_CREATE_STATE(e, r, x, p)** is the same as **HSM_CREATE_MSG_STATE** except
+no Entry or Exit messages are printed.
+
+**FSM_CREATE_MSG_STATE(me, mx, e, r, x)** is used to create a flat state and
+prints a message in the optional Entry action and a message in the optional
+Exit action. The macro has the following parameters:
+
+    - me - String message that's printed in the Entry action. Can be NULL.
+    - mx - String message that's printed in the Exit action. Can be NULL.
+    - e - Function pointer to state's Entry action. Can be NULL.
+    - r - Function pointer to state's Run action. Can be NULL.
+    - x - Function pointer to state's Exit action. Can be NULL.
+
+**FSM_CREATE_STATE(e, r, x)** is the same as **FSM_CREATE_MSG_STATE** except no
+Entry or Exit messages are printed.
+
+To enable message printing, the **CONFIG_SMF_PRINT_MSG=y** config option must be
+added to the projects **prj.conf** file. The use of **HSM_CREATE_MSG_STATE** or
+**FSM_CREATE_MSG_STATE** when the **CONFIG_SMF_PRINT_MSG** config option isn't
+defined, results in a build error.
+
+To enable a flat state machine, the **CONFIG_SMF_FLAT=y**  config option must be
+added to the projects **prj.conf** file. The use of **FSM_CREATE_MSG_STATE** or
+**FSM_CREATE_STATE** when the **CONFIG_SMF_FLAT** config option isn't defined,
+results in a build error. If the **CONFIG_SMF_FLAT=y** config option is defined
+and **HSM_CREATE_STATE** or **HSM_CREATE_MSG_STATE** is used, results in a build
+error.
+
+State Machine Creation
+======================
+
+A state machine is created by defining a table of states that's indexed by an
+enum. For example::
+
+	enum demo_state { S0, S1, S2 };
+
+	const struct smf_state demo_states {
+		[S0] = FSM_CREATE_STATE(s0_entry, s0_run, s0_exit),
+		[S1] = FSM_CREATE_STATE(s1_entry, s1_run, s1_exit),
+		[S2] = FSM_CREATE_STATE(s2_entry, s2_run, s2_exit)
+	};
+
+To set the initial state or to transition from one state to another, the
+**set_state** function is used and it has the following prototype:
+**void set_state(smf_ctx *ctx, smf_state *state)**
+
+**NOTE:** While the state machine is running, set_state should only be called
+from the Entry and Run functions. Calling set_state from the Exit functions
+doesn't make sense and will generate a warning.
+
+State Machine Execution
+=======================
+To run the state machine, the **run_state** function should be called in some
+application dependent way. The function has the following prototype:
+**void run_state(smf_ctx *ctx)**
+
+Flat State Machine Example
+==========================
+This example turns the following state diagram into code using the SMF, where
+the initial state is S0.
+
+.. figure:: img/flat.png
```

Comment:
note: you may want to use `graphviz` diagrams, they are natively supported in the documentation (render to SVG and are better viewed in both light/dark modes)

Reply:
Yes, the image is much better. Thanks.